### PR TITLE
fix issue 19729 - Constructor overloads coming from mixin are not resolved

### PR DESCRIPTION
**Pull request by ghost**
The test _b17259.d_ was created to check that an ICE was fixed but then this became a reject-valid.
Make it working for good this time.

### CODE REVIEW

**Review by thewilsonator — APPROVED** [review]

File: src/dmd/func.d
Comment (Geod24) [inline src/dmd/func.d:2953]: ?

File: src/dmd/func.d
Comment (Geod24) [inline src/dmd/func.d:2955]: Please use a link

**Review by Geod24 — CHANGES_REQUESTED** [review]
This needs a bit more test.

1. Add a test with named mixin
2. How does it work in case of conflict ?
3. How does it work if there's an exact match mixed in and a conversion match at the top level

File: src/dmd/func.d
Comment (NilsLankila) [reply src/dmd/func.d:2953]: Hard to say if this still happens.

Comment (MoonlightSentinel) [reply src/dmd/func.d:2953]: Seems like it's never hit by the test suite:

https://codecov.io/gh/dlang/dmd/compare/fe263f5991ea571b0384c6facfe207a7b17efb44...c3f90f86d8156d2fcb8d10e48add707fb6e6072e/diff

**Comment by NilsLankila** [comment]
I can answer, as overload sets are tried at the end:

> How does it work in case of conflict ?

_mixin looses, top level decl wins_

> How does it work if there's an exact match mixed in and a conversion match at the top level

_mixin looses, top level decl wins_

But I'll push tests.

File: test/runnable/test19729.d
Comment (MoonlightSentinel) [inline test/runnable/test19729.d:1]: Please add
```
PERMUTE_ARGS:
```
there's no need to run permutations for this test

File: src/dmd/func.d
Comment (NilsLankila) [reply src/dmd/func.d]: I  thought to put `fatal()` instead of returning null actually. If by accident someone manages to trigger the new explicit ICE then there's something to fix anyway.

**Comment by NilsLankila** [comment]
ping @Geod24, I have pushed the tests you required four days ago, requested a second review but you still block this bugfix.

File: test/fail_compilation/fail19729.d
Comment (wilzbach) [inline test/fail_compilation/fail19729.d:16]: FYI: there's no requirement to format the tests and even in general `this(T t) {}` would be allowed ;-)

File: test/runnable/test19729.d
Comment (wilzbach) [inline test/runnable/test19729.d:2]: Should have an empty `TEST_OUTPUT` to ensure no silent deprecations or other messages are triggered.

```
TEST_OUTPUT:
---
---
```

Or are we adding this by default now?

Comment (MoonlightSentinel) [reply test/runnable/test19729.d:2]: Yes, empty `TEST_OUTPUT` is the default nowadays.

File: test/fail_compilation/fail19729.d
Comment (Geod24) [inline test/fail_compilation/fail19729.d:23]: Please also add:
```D
// Those two types will be ambiguous with `string` as an argument
class D
{
    mixin Templ!(const(char)*);
    mixin Templ!(const(char)[]);
}

// ...
new D("conflict");
```

**Review by Geod24 — APPROVED** [review]
> ping @Geod24, I have pushed the tests you required four days ago, requested a second review but you still block this bugfix.

Not on purpose, I promise, because I'm [looking forward to this bug fix too](https://github.com/Geod24/minivariant/blob/d20b240d45289a857b4c0d6e7f842a92b3d4a19e/source/geod24/variant.d#L270-L281) :)
For reference, anyone is free to discard an (addressed) review if the reviewer is busy.

I added a last test case, otherwise LGTM.

**Pull request merged**
